refactor: general suppressions (#2302)
biomejs/website: src/content/docs/analyzer/suppressions.mdx

---
title: Suppressions
description: How to use suppressions.
---

> The Biome analyzer is the foundation of the [linter](/linter) and [assist;](/assist) in fact, both tools share a lot of similarities.

Reviewer: ```suggestion
The Biome analyzer is the foundation of the [linter](/linter) and [assist](/assist); in fact, both tools share a lot of similarities.
```

Among them, they share the same suppression engine, which means that you can suppress a lint rule the same way you would suppress an assist action.

> With suppression, it's possible to turn off a lint rule (or action) for a specific line of the code, more lines of code, or the entire file.

Reviewer: ```suggestion
With suppression, it's possible to turn off a lint rule (or action) for a specific line of code, a range, or the entire file.
```

Suppression is achievable with **suppression comments**.

## Suppression syntax

Suppression comments have the following format:

```js
// biome-ignore lint: <explanation>
// biome-ignore assist: <explanation>
// biome-ignore syntax: <explanation>
// biome-ignore lint/suspicious: <explanation>
// biome-ignore lint/suspicious/noDebugger: <explanation>
// biome-ignore lint/suspicious/noDebugger(foo): <explanation>
// biome-ignore-all lint: <explanation>
// biome-ignore-start lint: <explanation>
// biome-ignore-end lint: <explanation>
```

> Let's break it down:

Reviewer: The part that follows felt a bit off to me, since we're not actually breaking down a single example. So I've rephrased a bit below.

- `biome-ignore`, `biome-ignore-all`, `biome-ignore-start` and `biome-ignore-end` are the start of a suppression comment.

> - `lint` is the **category** of the suppression comment, and in this case it suppresses the linter. It can be `lint`, `action` or `syntax`.

Reviewer: ```suggestion
- After the space follows the **category** of the suppression comment. It can be `lint`, `assist`, or `syntax`.
```

> - `/suspicious/noDebugger` or `/suspicious`: **optional**, group and name, or group of the rule you want to suppress.

Reviewer: ```suggestion
- The category is followed by an optional group, or group and name, separated by slashes. For example, `/suspicious` or `/suspicious/noDebugger`. The more you specify, the more specific (i.e. targeted) your suppression is.
```

> - `(foo)`: **optional**, values associated to the suppression comment, these values are used by the specific rules. Refer to the documentation

Reviewer: ```suggestion
- Some rules even allow you to specify values during suppression. These can be specified between parentheses, for example: `(foo)`. Refer to the rule documentation to learn more about these.
```

of the single rule to learn if they are used.
- `<explanation>` explanation why the rule is disabled.

> If you're unsure what's the exact category of a rule/action, you can refer to their documentation page and use their *diagnostic category*.

Reviewer: ```suggestion
If you're unsure of the exact category of a rule/action, you can refer to their documentation page and use their *diagnostic category*.
```

:::note
From here onwards, JavaScript is used to explain suppression comments, and the `lint` category is used.
:::

### Inline suppressions

They disable a lint rule for the **next line** of code.

In the following example, the suppression comment `biome-ignore lint/suspicious/noDebugger: reason` will disable the `debugger;` statement at line 2, but the `debugger` at line 3 will still raise a diagnostic:

```js title="file.js" ins={1}
// biome-ignore lint/suspicious/noDebugger: reason
debugger;
debugger;
```

### Top-level suppressions

They disable a lint rule for an entire file. They must be placed **at the top of the file**, and they must start with `biome-ignore-all`.

> These suppression comments are very useful when you want to suppression some lint rule *for a particular file*, and you don't want to rely to a single configuration override to achieve that.

Reviewer: ```suggestion
These suppression comments are very useful when you want to suppression some lint rule *for a particular file*, and you don't want to rely on a single configuration override to achieve that.
```

> The following example, the suppression comment `biome-ignore-all lint/suspicious/noDebugger: reason` will disable the lint rule for the whole file `generated.js`:

Reviewer: ```suggestion
In the following example, the suppression comment `biome-ignore-all lint/suspicious/noDebugger: reason` will disable the lint rule for all lines in `generated.js`:
```

```js title="generated.js" ins={1}
// biome-ignore-all lint/suspicious/noDebugger: reason

debugger
debugger

```

When a top-level suppression comment isn't at the top of the file, it is considered *unused* and Biome will emit a diagnostic with category `suppression/unused`.

### Range suppressions

> They disable a lint rule from a particular line of code, to another line of code, called range.

Reviewer: ```suggestion
They disable a lint rule from a particular _range_ in the file, starting from the line with the start comment, until the line with the end comment.
```

> To mark the beginning of range suppression, the suppression comment must start with `// biome-ignore-start`, and to mark the end of it, the suppression comment must start with `// biome-ignore-end`.

Reviewer: ```suggestion
To mark the beginning of a range suppression, the suppression comment must start with `// biome-ignore-start`. To mark the end of it, the suppression comment must start with `// biome-ignore-end`.
```
